File: Python DS & Algo - Runestone Academy/4. Basic Data Structures/4.27.x_queue_using_linkedlist.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class Queue:
    def __init__(self):
        self.front = None
        self.rear = None

    def isEmpty(self):
        return self.front == None

    def enqueue(self, item):
        newNode = Node(item)
        if self.front == None: # queue is empty
            self.front = self.rear = newNode
        else:
            self.rear.next = newNode
            self.rear = self.rear.next

    def dequeue(self):
        if self.isEmpty():
            return 'Queue is empty!'
        return_num = self.front.data
        self.front = self.front.next
        return return_num

    def display(self):
        current = self.front
        print('F -> ', end = '')
        while current != None:
            print(current.data, end = ' ')
            current = current.next
        print('<- R', end = '')

q = Queue()
q.enqueue(1)
q.enqueue(2)
q.enqueue(3)
q.dequeue()
q.dequeue()
q.enqueue(4)
q.enqueue(5)
q.display()

File: Python DS & Algo - Runestone Academy/4. Basic Data Structures/C++/4.27.x_queue_using_linked_list.cpp
#include<iostream>
using namespace std;

struct Node{
    int data;
    struct Node * next = NULL;
};

class Queue{
private:
    struct Node * front = NULL;
    struct Node * rear = NULL;
public:
    void enqueue(int data){
        struct Node * newNode = new Node;
        newNode -> data = data;
        if(front == NULL)
            front = rear = newNode;
        else{
            rear -> next = newNode;
            rear = rear -> next;
        }
    }

    int dequeue(){
        if(front == NULL)
            return -1;
        else{
            int return_num = front -> data;
            front = front -> next;
            return return_num;
        }
    }

    void display(){
        cout << "F -> ";
        Node * current = front;
        while(current){
            cout << current -> data << " ";
            current = current -> next;
        }
        cout << "<- R " << endl;
    }
};

int main(void){
    Queue q;
    q.enqueue(1);
    q.enqueue(2);
    q.enqueue(3);
    q.dequeue();
    q.dequeue();
    q.enqueue(4);
    q.enqueue(5);
    q.display();
}
